### ISSUE TITLE
fetch client endpoints are typed as `any` on Typescript `5.5.x`

### ISSUE THREAD

**Issue by guillaume-docquier-vention** [ts-rest/ts-rest#647]
### Describe the bug

When creating a client from my contract on a NodeJS server using Typescript `5.5.x` (tested with `.2`, `.3` and `.4`), the endpoints are typed as `any`, losing all type information.
My IDE reports that type and `tsc` as well.

![image](https://github.com/user-attachments/assets/580862f8-ce12-48da-af76-ed17e2fefa50)

It works fine on `5.4.5`

### How to reproduce

Edit: clone this repo: https://github.com/guillaume-docquier-vention/ts-rest-typescript-5.5 and follow the readme

- Setup a NodeJS project using Typescript `5.5.x`
- Create a contract
- Create a client from the contract, using `initClient`
- Use any endpoint from the client

Notice that the endpoints are not typed.

### Expected behavior

All endpoints should be typed.
or
The docs/package should say that it is not compatible with Typescript `5.5.x`

### Code reproduction

```ts
import { initContract, initClient } from "@ts-rest/core"
import { z } from "zod"

export const contract = initContract().router(
  {
    auth: {
      login: {
        method: "POST",
        path: `/auth/login`,
        body: z.object({
          email: z.string(),
          password: z.string(),
        }),
        responses: {
          200: z.object({
            token: z.string(),
          }),
        },
      },
    },
    logout: {
      method: "POST",
      path: `/logout`,
      body: z.object({
        email: z.string(),
        password: z.string(),
      }),
      responses: {
        200: z.object({
          token: z.string(),
        }),
      },
    },
  },
  {
    strictStatusCodes: true,
    pathPrefix: "/api",
  },
)

const client = initClient(contract, {
  baseUrl: `http://localhost:3000`,
  baseHeaders: {},
})

client.auth.login("invalid") // Problem: type of "login" is "any"
client.logout("invalid") // Problem: type of "logout" is "any"
client.invalid() // As expected: TS2339: Property invalid does not exist on type
```

### ts-rest version

3.46.0

### Other versions

Node 22.3.0
Npm 10.8.1
Typescript 5.5.2 / 5.5.3 / 5.5.4

**Comment by guillaume-docquier-vention** [comment]
For now I'll just downgrade to `5.4.5` since I don't really need `5.5.x`, but I thought it would be good to be aware of this.

**Comment by Gabrola** [comment]
I would say this is a duplicate of #618, and is an IDE error, not ts-rest or Typescript, but it's weird that `tsc` is not erroring for you.

You can see [here](https://www.typescriptlang.org/play/?strict=false&strictNullChecks=false&ts=5.5.3#code/JYWwDg9gTgLgBAbzsAdsGBhCKZQIYDGMANMmpgDbACmOcAvnAGZQQhwDkAAjAM4C0UarxgB6AtGocA3AChQkWIjgAvBs1bsOKiABMZs2dQAei+BJQi4F3IXgBeMuiw58RABQBKAHSsArjDUUO6ycIihYXB4AQAWAFzhkZEUEADmqAkIEUlhINQwMXoJHAAKAPIAygAqHMTZOWB4BQkABqLRBaIp6SgtdTmRAEZ6AJ4JKt4QgwBW1B5ZA0nUIHjAFOPeIlCoqV79i2GNvLwA7tC6G1s7e-VJ9J77A0K8kJbCmbc5AEwADD8bU1m80+AxgEAA1rRLrhrg8QZF7o8cvQkWEUfV0UluhAAh8BnkCkVOOVqrVPo1mnA2tiAn1PsNdGNVJMZnMYO4Fotlqt1syrihdnCDkdTudodsBTcBojPs9Xrx3olFr9-szAWyOfCwmDIShxbDUQihcikZiGPtOWErkQKjAmn5eFhdIrcH5qEiKTESkImMBjMV2mBgGS0XVPHJZBYrAQqLQHE5KDQcO4bG4SErBngFQBVKC8loxGAwMBxURdCAEPAUQoiOIAZj+PzpYUzCoAEtQ8M6oLxMuj7hGY0mYN4OjFvN1UO4OKgAG5V4D6cOR2M4CdpHHsmcoedUJdyIdx7xzhe6LxyIA), and in the output below that that the types are evaluated correctly, and erroring with invalid input.

```shell
$ tsc -v
Version 5.5.4
$ tsc
src/test.ts:46:19 - error TS2559: Type '"invalid"' has no properties in common with type '{ cache?: RequestCache; body?: { email?: string; password?: string; }; fetchOptions?: FetchOptions; extraHeaders?: Record<string, string>; overrideClientOptions?: Partial<...>; }'.

46 client.auth.login('invalid');
                     ~~~~~~~~~

src/test.ts:47:15 - error TS2559: Type '"invalid"' has no properties in common with type '{ cache?: RequestCache; body?: { email?: string; password?: string; }; fetchOptions?: FetchOptions; extraHeaders?: Record<string, string>; overrideClientOptions?: Partial<...>; }'.

47 client.logout('invalid');
                 ~~~~~~~~~

src/test.ts:48:8 - error TS2339: Property 'invalid' does not exist on type 'RecursiveProxyObj<RecursivelyApplyOptions<{ auth: { login: { method: "POST"; path: "/auth/login"; body: ZodObject<{ email: ZodString; password: ZodString; }, "strip", ZodTypeAny, { email?: string; password?: string; }, { ...; }>; responses: { ...; }; }; }; logout: { ...; }; }, { ...; }>, { ...; }>'.

48 client.invalid();
          ~~~~~~~
```

Please submit a link to a full repo with this problem.

Disregarding the `tsc` issue. IDEs will very often take a long time to compute or infer complex types, so while this is happening in the background the type will show as `any` and you would not get any errors, red line, or completion until that happens. Even after the types are determined, complex types might still show as `any` 

I'm curious, because this might be related to your other random autocompletion issues. Do you have this "[Use types from server](https://www.jetbrains.com/help/webstorm/typescript-support.html#ws_ts_restart_ts_service)" option turned on in WebStorm? I tried it and noticed it was a bit unstable and very often super slow or even fails to compute types sometimes.

**Comment by guillaume-docquier-vention** [comment]
If your tsc report errors correctly, would that mean I have a wrong tsconfig, maybe?
I'll setup a repro repository.

I do use `use types from server`, but I also tried without and nothing changed.

**Comment by guillaume-docquier-vention** [comment]
Here's the repro: https://github.com/guillaume-docquier-vention/ts-rest-typescript-5.5

**Comment by Gabrola** [comment]
@guillaume-docquier-vention you need to add `"dom"` to your `compilerOptions.lib`

**Comment by guillaume-docquier-vention** [comment]
Why should I add "dom" if I'm using the client server side?

**Comment by Gabrola** [comment]
@guillaume-docquier-vention the types for `fetch` and `AbortSignal` are in `dom`

**Comment by guillaume-docquier-vention** [comment]
`@types/node` provides the `fetch` and `AbortSignal` types, why do I specifically need `dom`? I'm mostly asking because I would rather not have `dom` in my NodeJS server.

(I've updated the repro with `@types/node` to show that the types are there)

I'm assuming ts-rest uses the `dom` fetch type, which is why it doesn't work without it. I don't know if there's something we could do to get platform agnostic types? Something like [an isomorphic library](https://stackoverflow.com/questions/65848540/right-way-to-reference-platform-specific-types-in-isomorphic-typescript-library) or the good old [`setTimeout` type trick](https://stackoverflow.com/questions/45802988/typescript-use-correct-version-of-settimeout-node-vs-window). It might not be very straightforward, but it would be nice.

Or did you not intent the client to be used server side (although that seems useful to me)?
In the meantime, of course, I'll use `dom` but as I said, it feels awkward in a NodeJS project.

**Comment by Gabrola** [comment]
You will not have `dom` on your server. `dom` is just some type declarations a built into typescript. You are only making the types visible to the typescript compiler by adding it your `tsconfig.json`, but there is absolutely zero effect on your runtime.

More importantly, until very recently, Node support for `fetch` was not good (`fetch` only became officially stable in Node v21). Regardless, `@types/node` v18 is missing most of the types used by `@ts-rest/core`, and now that I checked, even v20 is missing `RequestCredentials` and `RequestCache`, and they probably never will be added to node, because these are related to cookie CORS rules, and browser caching, both irrelevant to a node environment. These are the types we need, and you can check for yourself that they do not exist.

```typescript
type T1 = RequestInit;
type T2 = RequestCredentials;
type T3 = FormData;
type T4 = URLSearchParams;
type T5 = AbortSignal;
type T6 = RequestCache;
type T7 = Headers;
type T8 = typeof fetch;
type T9 = File;
```

Either way, built-in TS `lib`s become available globally to all packages when you add it to your project. It's not the same for `@types/*` packages. I'd have to add it to the dependencies of `@ts-rest/core` if you wanted to use that instead. Which is obviously not ideal when the client is mainly used in the browser. Also makes no sense to add a dependency when the types are already available built-in to typescript.

**Comment by guillaume-docquier-vention** [comment]
I know it only adds types, not runtime stuff. It adds types for a lot of browser things aside from fetch, and allows me to use things like `document` and `window` without errors.

Of course we can be reponsible and just know that `window` and `document` don't exist Node, we don't have a reason to use them, but:
- I don't trust that everyone that works on the project will be responsible
- It can get confusing for types that are in both NodeJS and the Browser, like `Event` and `EventTarget` (I don't know if they have the same shape, but [they don't behave exactly the same](https://nodejs.org/api/events.html#nodejs-eventtarget-vs-dom-eventtarget))
- Fundamentally, I just don't want types for things that I can't use

But for me the most important thing is
> v20 is missing RequestCredentials and RequestCache, and they probably never will be added to node, because these are related to cookie CORS rules, and browser caching, both irrelevant to a node environment.

If you are using these in the client, will it work in node?

**Comment by guillaume-docquier-vention** [comment]
And also, why does it work on TS 5.4 and not 5.5?

**Comment by Gabrola** [comment]
> If you are using these in the client, will it work in node?

The client, yes, will work on node, but it will just ignore these options. However, it's only just relatively recently that node started supporting `fetch`. We never really made a deliberate effort to make the client work on node, simply because there was no `fetch` pre-v18, and even file uploading was not working between 18.0  and 18.12 because there was no support for the `File` object. However, `fetch` and other interfaces like `File` could be polyfilled, so there was really no reason really to put any extra effort towards removing any dependency on `fetch`.

> And also, why does it work without `dom` on TS 5.4 and not 5.5?

There's something weird with TS 5.5 when it comes to missing types. You can see an extremely simplified example [here](https://www.typescriptlang.org/play/?lib=lib.es2023.d.ts&ts=5.5.3#code/PTAEAEBsEsCMC5QFMDOAmADGgzAKFwC4CeADkqAEICGK0AxgEpICOArqgaALygDeuoQUJCgAegH4BoWAHsAJkUQoCAJ2gA7AOYBuKXSp0AFkkRM2HAMIHjugL75iZUAHEkBANJIiASQBmABRUZMhViABVSJAB5FQA5JAA3JBUAHjCAGlB3ZAAPAiR1ORRQAGsvGV9QDNAANQA+blAAURy6SFY5JDSAbXcAXUzWQqRfDSQ5BqQ8gqLa0HEs0ER1ROTdQkisr0bXDy8-QODk8MiY+KTU6lpGFnZlTIByfSMkB8zlNS063SFQEQkpIIRARDNBiihDDJWJA5NJyE9rK9QDIVKAVhcgA) of a part of our type codes that we use internally. We should expect the type to be either the `K` we pass, which is `cache` in this case, or `never` but somehow it returns `any`. This cascades further into our other types, and we end up with the client functions typed as `any`. You can see on the playground that the type turns into `cache` if you set the version to TS 5.4 🤷‍♂️

I think it's part of the optimizations they made so TS doesn't waste too much time evaluating supposed dead-ends such as this one where the type is invalid/missing.

**Comment by guillaume-docquier-vention** [comment]
I see, thank you!
You know your stuff!

**Comment by Gabrola** [comment]
@guillaume-docquier-vention I found a workaround for this issue. Should work now in TS 5.5 even if you don't have `dom` in your libs.

**Comment by guillaume-docquier-vention** [comment]
This is great news! Would you mind sharing what the workaround is? At a high level

**Comment by bestickley** [comment]
I also just ran into this issue. @guillaume-docquier-vention, thank you for reporting! saved me lots of time to read this.
@Gabrola, in my experience, when i need fetch and AbortSignal types, I've used: https://www.npmjs.com/package/undici-types. They're compliant with Node.js fetch. I think there is relation to @types/node but not 100% sure. I wish fetch was included in @types/node. Looks like this is resolved but just wanted to share FYI.

**Comment by Gabrola** [comment]
@bestickley `undici-types` is a dependency of `@types/node` globals AFAIK. `undici-types` by itself if included as a type dependency in public packages causes mayhem if you use `lib.dom` in a consuming app/lib, since you cannot reliably pass any `lib.dom` typed data to any function that is typed using `undici-types`. There is no guarantee that the types do not diverge between `undici-types` and `lib.dom`.

I just realized though that `@types/node` [conditionally sets the types](https://github.com/andrewbranch/DefinitelyTyped/blob/124c5f37d0e894d6fd10f832ffc4dd426a2c0fa5/types/node/globals.d.ts#L6) based on if `lib.dom` is included or not. I might copy this into our code honestly.

**Comment by bestickley** [comment]
@Gabrola, you're right. I'm very surprised to see `undici-types` [is dependency of](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/package.json#L12) `@types/node`. I did a deep dive on this a year ago now and I thought that wasn't the case. If `@types/node` includes `undici-types` which has `fetch` and friends, why was `DOM` required for this feature to work?

>  in public packages causes mayhem

I was unaware. Good to know.

I spoke too soon, I just ran into this issue with latest ts-rest (^3.49.3). When I enabled `DOM` I get type errors (correct behavior) but when I remove `DOM` (which I want to because I'm in Node.js environment) then it is `any` and I don't get type errors (incorrect behavior) :/

Do you want me to create new issue or re-open this one? I'm in PNPM Monorepo using [internal package](https://turbo.build/repo/docs/crafting-your-repository/creating-an-internal-package) pattern. Maybe that has something to do with it? Clearly getting behavior describe in this issue.

**Comment by guillaume-docquier-vention** [comment]
@bestickley I did not validate that the issue was resolved by the latest version(s) of ts-rest, I'll find some time to update my repro, maybe that can help you debug on your side

**Comment by Gabrola** [comment]
@bestickley Ah, I just checked, and it seems like the fix resolved only an internal type resolving to `any`, but seems like `any` is bubbling up from somewhere else. Anyway, I don't want to waste too much time trying to fix this specific TS 5.5 issue. I'll just use conditional types instead.
